### PR TITLE
update indirect_structural_match and pointer_structural_match lints to match RFC and to show up for dependencies

### LINKED ISSUE #73448: Handle generic associated constants in const qualification for structural match
#67343 started using const qualification for more precise structural match warnings. The idea is that const qualification should be strictly more precise than the current type visitor (`search_for_structural_match_violation`). However, const qualification is more conservative in the face of generic associated constants than the type visitor, which typically runs after substitution. This was the case in the example in #73431.

We will need to make const qualification handle all cases that `search_for_structural_match_violation` does before switching away from the latter.

### PR DESCRIPTION
**Pull request by RalfJung**
This is a large step towards implementing https://github.com/rust-lang/rfcs/pull/3535.
We currently have five lints related to "the structural match situation":
- nontrivial_structural_match
- indirect_structural_match
- pointer_structural_match
- const_patterns_without_partial_eq
- illegal_floating_point_literal_pattern

This PR concerns the first 3 of them. (The 4th already is set up to show for dependencies, and the 5th is removed by https://github.com/rust-lang/rust/pull/116284.) nontrivial_structural_match is being removed as per the RFC; the other two are enabled to show up in dependencies.

Fixes https://github.com/rust-lang/rust/issues/73448 by removing the affected analysis.

### CODE REVIEW

File: tests/ui/consts/const_in_pattern/accept_corner_cases.rs
Comment (RalfJung) [inline tests/ui/consts/const_in_pattern/accept_corner_cases.rs:1]: Accepting these patterns is the key change here -- we no longer plan to reject these in the future so we don't have to lint against them.

**Comment by traviscross** [comment]
@rustbot labels +T-lang +I-lang-nominated

This is part of implementing rust-lang/rfcs#3535, which we accepted.  Let's nominate to verify we're OK to stabilize this piece.

@RalfJung: If there's anything beyond what is in the PR description you'd like the team to understand when considering this piece, it may be worth posting a stabilization report to that effect.

**Comment by RalfJung** [comment]
The nontrivial_structural_match is being removed, meaning we now accept code like this without a lint (where `NoDerive` is a type with a custom `PartialEq`):
```rust
    // The final value is fine, but statically analyzing the expression that computes
    // the value would likely (incorrectly) have us conclude that this may match on
    // values that do not have structural equality.
    const INDEX: Option<NoDerive> = [None, Some(NoDerive(10))][0];
    match None { Some(_) => panic!("whoops"), INDEX => dbg!(INDEX), };

    const fn build() -> Option<NoDerive> { None }
    const CALL: Option<NoDerive> = build();
    match None { Some(_) => panic!("whoops"), CALL => dbg!(CALL), };

    impl NoDerive { const fn none() -> Option<NoDerive> { None } }
    const METHOD_CALL: Option<NoDerive> = NoDerive::none();
    match None { Some(_) => panic!("whoops"), METHOD_CALL => dbg!(METHOD_CALL), };
```

The indirect_structural_match and pointer_structural_match lints detect constants that do not have structural equality according to the definition from the RFC, but are currently accepted:
- indirect_structural_match detects constants involving references: due to a bug, `&NotStructural`  values (and more generally, having a type with custom equality behind a reference) was not detected properly. We have a future-compat lint against this for many years already. This makes the lint show up in dependencies as well, in preparation for making it a hard error.
- pointer_structural_match detects constants containing raw pointers and function pointers that are not just bare integers.  According to the RFC, such values are not considered to have structural equality. The crater run in https://github.com/rust-lang/rust/pull/116930 found no case of the lint actually triggering. The lint is warn-by-default since Rust 1.75; this makes the lint show up in dependencies as well.

**Comment by traviscross** [comment]
@rustbot labels -I-lang-nominated -S-waiting-on-team

We discussed this in the triage meeting today with all members present and agreed that we want to do this.  This is OK to proceed.

Thanks again to @RalfJung for pushing forward on this.

**Pull request merged**